sanic/server.py

```diff
@@ -147,12 +149,29 @@ def on_message_complete(self):
     # -------------------------------------------- #
 
     def write_response(self, response):
-        try:
-            keep_alive = self.parser.should_keep_alive() \
-                            and not self.signal.stopped
+
+        def attempt_write(_response):
             self.transport.write(
-                response.output(
+                _response.output(
                     self.request.version, keep_alive, self.request_timeout))
+
+        try:
+            keep_alive = (self.parser.should_keep_alive()
+                          and not self.signal.stopped)
+
+            try:
+                # If we already have an HTTPResponse object this should be
+                # as fast as ever
+                attempt_write(response)
+            except AttributeError:
+                try:
+                    # A performant way to check if we have a list or dict
+                    # Both list and dict contain a `clear` function
+                    response.clear
+                    attempt_write(json(response))
```

Comment:
This would be cleaner and theoretically more performant if it used a type map instead of the doubly nested try/except:

```python
TYPE_MAP = {
    str: text,
    dict: json
     }

try:
    attempt_write(response)
except AttributeError:
    try:
        attempt_write(TYPE_MAP(response))
    except Exception as e:
        self.bail_out("Writing response failed,\
                      connection closed {}".format(e))
```